invoices/invoiceregistry.go

```diff
@@ -465,19 +598,95 @@ func (i *InvoiceRegistry) NotifyExitHopHtlc(rHash lntypes.Hash,
-		// checking to do.
-		if inv.Terms.State == channeldb.ContractCanceled {
-			debugLog("invoice already canceled")
-			return nil, errNoUpdate
+		// Start building the accept descriptor.
+		acceptDesc := &channeldb.HtlcAcceptDesc{
+			Amt:          amtPaid,
+			Expiry:       expiry,
+			AcceptHeight: currentHeight,
 		}
 
-		// If an invoice amount is specified, check that enough
-		// is paid. Also check this for duplicate payments if
-		// the invoice is already settled or accepted.
-		if inv.Terms.Value > 0 && amtPaid < inv.Terms.Value {
-			debugLog("amount too low")
-			return nil, errNoUpdate
+		// If an invoice amount is specified and this is not an mpp,
+		// check that the exact amount is paid.
+		var canSettle bool
+		if mpp == nil {
+			// If the invoice is already canceled, there is no further
+			// checking to do.
+			if inv.Terms.State == channeldb.ContractCanceled {
+				debugLog("invoice already canceled")
+				return nil, errNoUpdate
+			}
+
+			// If an invoice amount is specified, check that enough
+			// is paid. Also check this for duplicate payments if
+			// the invoice is already settled or accepted.
+			if inv.Terms.Value > 0 && amtPaid < inv.Terms.Value {
+				debugLog("amount too low and no mpp")
+				return nil, errNoUpdate
+			}
+
+			// TODO(joostjager): Check invoice mpp required feature
+			// bit.
+
+			// Don't allow settling the invoice with an old style
+			// htlc if we are already in the process of gathering an
+			// mpp set.
+			for _, htlc := range inv.Htlcs {
+				if htlc.TotalAmt > 0 {
+					debugLog("mpp reception in progress")
+					return nil, errNoUpdate
+				}
+			}
+
+			canSettle = true
+		} else {
+			// Only accept payments to open invoices.
+			if inv.Terms.State != channeldb.ContractOpen {
+				debugLog(fmt.Sprintf(
+					"invoice in state %v and no longer open",
+					inv.Terms.State))
+				return nil, errNoUpdate
+			}
+
+			// Check the payment address that authorizes the
+			// payment.
+			if mpp.PaymentAddr() != inv.PaymentAddr {
+				debugLog("payment address mismatch")
+
+				return nil, errNoUpdate
+			}
+
+			// Check whether total amt matches other htlcs in the
+			// set.
+			for _, htlc := range inv.Htlcs {
+				// Only consider accepted mpp htlcs.
+				if htlc.State != channeldb.HtlcStateAccepted {
+					continue
+				}
+
+				if mpp.TotalMsat() != htlc.TotalAmt {
+					debugLog("htlc total amt doesn't " +
+						"match set total")
+					return nil, errNoUpdate
+				}
+			}
+
+			// If there is a set invoice amount, check that the
+			// total amt of the htlc set is high enough.
+			if inv.Terms.Value > 0 {
+				if mpp.TotalMsat() < inv.Terms.Value {
+					debugLog("set total too low")
+					return nil, errNoUpdate
+				}
+			}
+
+			// Make sure the communicated set total isn't overpaid.
+			if inv.AmtPaid+amtPaid > mpp.TotalMsat() {
+				debugLog("mpp is overpaying set total")
+				return nil, errNoUpdate
+			}
+
+			canSettle = inv.AmtPaid+amtPaid == mpp.TotalMsat()
```

Comment:
is the `inv.AmtPaid` updated as each htlc is accepted? should it not pull this from the htlc set directly, and only update after accept, so that the payment appears atomic externally?

Reply:
Good point. This triggered quite a refactor that is also good for the overall code quality of invoice registry.

Reply:
Has been merged as part of #3770